### ISSUE TITLE
Loss is not converging during training DPO 

### ISSUE THREAD

**Issue by yunsangju** [eric-mitchell/direct-preference-optimization#20]
Hello.
I already have an SFT model, so I only conducted DPO training.
The train loss doesn't seem to converge as shown below.
Is the normal situation correct?
![image](https://github.com/eric-mitchell/direct-preference-optimization/assets/136311766/864fc276-0e85-4600-a8e9-e058b13360d9)

below is our train code
- DPO Module
```python
class DPOModule(nn.Module):
    def __init__(self, train_config):
        super().__init__()
        dpo_config = train_config.dpo
        self.policy_model = AutoModelForCausalLM.from_pretrained(
            dpo_config.policy.path,
            low_cpu_mem_usage=True,
            torch_dtype=torch.float32,
        )
        self.reference_model = AutoModelForCausalLM.from_pretrained(
            dpo_config.reference.path,
            low_cpu_mem_usage=True,
            torch_dtype=torch.float16,
        )
        self.reference_model.eval()
        self.dpo_beta = dpo_config.beta

    def _dpo_loss(
        self,
        policy_chosen_logps: torch.FloatTensor,
        policy_rejected_logps: torch.FloatTensor,
        reference_chosen_logps: torch.FloatTensor,
        reference_rejected_logps: torch.FloatTensor,
        beta: float,
        reference_free: bool = False,
    ) -> Tuple[torch.FloatTensor, torch.FloatTensor, torch.FloatTensor]:
        """Compute the DPO loss for a batch of policy and reference model log probabilities.
        Args:
            policy_chosen_logps: Log probabilities of the policy model for the chosen responses. Shape: (batch_size,)
            policy_rejected_logps: Log probabilities of the policy model for the rejected responses. Shape: (batch_size,)
            reference_chosen_logps: Log probabilities of the reference model for the chosen responses. Shape: (batch_size,)
            reference_rejected_logps: Log probabilities of the reference model for the rejected responses. Shape: (batch_size,)
            beta: Temperature parameter for the DPO loss, typically something in the range of 0.1 to 0.5. We ignore the reference model as beta -> 0.
            reference_free: If True, we ignore the _provided_ reference model and implicitly use a reference model that assigns equal probability to all responses.
        Returns:
            A tuple of three tensors: (losses, chosen_rewards, rejected_rewards).
            The losses tensor contains the DPO loss for each example in the batch.
            The chosen_rewards and rejected_rewards tensors contain the rewards for the chosen and rejected responses, respectively.
        """
        pi_logratios = policy_chosen_logps - policy_rejected_logps
        ref_logratios = reference_chosen_logps - reference_rejected_logps

        if reference_free:
            ref_logratios = 0

        logits = pi_logratios - ref_logratios

        losses = -F.logsigmoid(beta * logits)
        chosen_rewards = (
            beta * (policy_chosen_logps - reference_chosen_logps).detach()
        )
        rejected_rewards = (
            beta * (policy_rejected_logps - reference_rejected_logps).detach()
        )

        return losses, chosen_rewards, rejected_rewards

    def pad_to_length(
        self,
        tensor: torch.Tensor,
        length: int,
        pad_value: Union[int, float],
        dim: int = -1,
    ) -> torch.Tensor:
        if tensor.size(dim) >= length:
            return tensor
        else:
            pad_size = list(tensor.shape)
            pad_size[dim] = length - tensor.size(dim)
            return torch.cat(
                [
                    tensor,
                    pad_value
                    * torch.ones(
                        *pad_size, dtype=tensor.dtype, device=tensor.device
                    ),
                ],
                dim=dim,
            )

    def _concatenated_inputs(
        self, batch: Dict[str, Union[List, torch.LongTensor]]
    ) -> Dict[str, torch.LongTensor]:
        """Concatenate the chosen and rejected inputs into a single tensor.
        Args:
            batch: A batch of data. Must contain the keys 'chosen_input_ids' and 'rejected_input_ids', which are tensors of shape (batch_size, sequence_length).
        Returns:
            A dictionary containing the concatenated inputs under the key 'concatenated_input_ids'.
        """
        max_length = max(
            batch["chosen_input_ids"].shape[1],
            batch["rejected_input_ids"].shape[1],
        )
        concatenated_batch = {}
        for k in batch:
            if k.startswith("chosen") and isinstance(batch[k], torch.Tensor):
                pad_value = -100 if "labels" in k else 0
                concatenated_key = k.replace("chosen", "concatenated")
                concatenated_batch[concatenated_key] = self.pad_to_length(
                    batch[k], max_length, pad_value=pad_value
                )
        for k in batch:
            if k.startswith("rejected") and isinstance(batch[k], torch.Tensor):
                pad_value = -100 if "labels" in k else 0
                concatenated_key = k.replace("rejected", "concatenated")
                concatenated_batch[concatenated_key] = torch.cat(
                    (
                        concatenated_batch[concatenated_key],
                        self.pad_to_length(
                            batch[k], max_length, pad_value=pad_value
                        ),
                    ),
                    dim=0,
                )
        return concatenated_batch

    def _get_batch_logps(
        self,
        logits: torch.FloatTensor,
        labels: torch.LongTensor,
        average_log_prob: bool = False,
    ) -> torch.FloatTensor:
        """Compute the log probabilities of the given labels under the given logits.
        Args:
            logits: Logits of the model (unnormalized). Shape: (batch_size, sequence_length, vocab_size)
            labels: Labels for which to compute the log probabilities. Label tokens with a value of -100 are ignored. Shape: (batch_size, sequence_length)
            average_log_prob: If True, return the average log probability per (non-masked) token. Otherwise, return the sum of the log probabilities of the (non-masked) tokens.
        Returns:
            A tensor of shape (batch_size,) containing the average/sum log probabilities of the given labels under the given logits.
        """
        assert logits.shape[:-1] == labels.shape

        labels = labels[:, 1:].clone()
        logits = logits[:, :-1, :]
        loss_mask = labels != -100

        # dummy token; we'll ignore the losses on these tokens later
        labels[labels == -100] = 0

        per_token_logps = torch.gather(
            logits.log_softmax(-1), dim=2, index=labels.unsqueeze(2)
        ).squeeze(2)

        if average_log_prob:
            return (per_token_logps * loss_mask).sum(-1) / loss_mask.sum(-1)
        else:
            return (per_token_logps * loss_mask).sum(-1)

    def concatenated_forward(
        self, model: nn.Module, batch: Dict[str, Union[List, torch.LongTensor]]
    ) -> Tuple[torch.FloatTensor, torch.FloatTensor]:
        """Run the given model on the given batch of inputs, concatenating the chosen and rejected inputs together.
        We do this to avoid doing two forward passes, because it's faster for FSDP.
        """
        concatenated_batch = self._concatenated_inputs(batch)
        all_logits = model(
            input_ids=concatenated_batch["concatenated_input_ids"],
            attention_mask=concatenated_batch["concatenated_attention_mask"],
        ).logits.to(torch.float32)
        all_logps = self._get_batch_logps(
            all_logits,
            concatenated_batch["concatenated_labels"],
            average_log_prob=False,
        )
        chosen_logps = all_logps[: batch["chosen_input_ids"].shape[0]]
        rejected_logps = all_logps[batch["chosen_input_ids"].shape[0] :]
        return chosen_logps, rejected_logps

    def forward(self, batch):
        policy_chosen_logps, policy_rejected_logps = self.concatenated_forward(
            self.policy_model, batch
        )
        with torch.no_grad():
            (
                reference_chosen_logps,
                reference_rejected_logps,
            ) = self.concatenated_forward(self.reference_model, batch)
        losses, chosen_rewards, rejected_rewards = self._dpo_loss(
            policy_chosen_logps,
            policy_rejected_logps,
            reference_chosen_logps,
            reference_rejected_logps,
            beta=self.dpo_beta,
        )
        reward_accuracies = (chosen_rewards > rejected_rewards).float()
        metrics = {
            "rewards_chosen": chosen_rewards.cpu().numpy().tolist(),
            "rewards_rejected": rejected_rewards.cpu().numpy().tolist(),
            "rewards_accuracy": reward_accuracies.cpu().numpy().tolist(),
            "rewards/margins": (chosen_rewards - rejected_rewards)
            .cpu()
            .numpy()
            .tolist(),
        }
        return losses.mean(), metrics
```

- DPO System
```python
class DPOSystem(pl.LightningModule):
    def __init__(self, cfg):
        super().__init__()
        self.cfg = cfg
        self.model = DPOModule(self.cfg.train)

    def configure_optimizers(self):
        optimizer = None
        train_cfg = self.cfg.train
        if train_cfg.optimizer == "adam":
            optimizer = torch.optim.Adam(
                self.model.policy_model.parameters(),
                lr=train_cfg.lr,
            )
        elif self.cfg.optimizer == "adamw":
            optimizer = torch.optim.AdamW(
                self.model.policy_model.parameters(),
                lr=train_cfg.lr,
            )

        scheduler = None

        if self.cfg.train.lr_scheduler == "linear":
            scheduler = {
                "scheduler": get_linear_schedule_with_warmup(
                    optimizer,
                    num_warmup_steps=train_cfg.warmup_steps,
                    num_training_steps=train_cfg.max_steps,
                ),
                "interval": "step",
            }

        return {"optimizer": optimizer, "lr_scheduler": scheduler}

    def training_step(self, batch: torch.Tensor, batch_idx: torch.Tensor):
        losses, metrics = self.model(batch)

        self.log(
            "dpo_loss", losses, on_step=True, on_epoch=False, prog_bar=True
        )

        return losses
```

- Dataset
```python
class DPODataset(Dataset):
    def __init__(self, cfg):
        self.cfg = cfg
        self.tokenizer = AutoTokenizer.from_pretrained(cfg.tokenizer.path)
        self.data = load_dataset()
        )

    def __getitem__(self, index):
        input_text_chosen, input_text_rejected = self.data[index]
        chosen_inputs = self.tokenizer(
            input_text_chosen,
            add_special_tokens=True,
            max_length=self.cfg.max_len,
            padding="max_length",
            truncation=True,
            return_tensors="pt",
        )
        rejected_inputs = self.tokenizer(
            input_text_rejected,
            add_special_tokens=True,
            max_length=self.cfg.max_len,
            padding="max_length",
            truncation=True,
            return_tensors="pt",
        )
        return {
            "chosen_input_ids": chosen_inputs["input_ids"].squeeze(0),
            "chosen_attention_mask": chosen_inputs["attention_mask"].squeeze(0),
            "chosen_labels": chosen_inputs["input_ids"].squeeze(0).clone(),
            "rejected_input_ids": rejected_inputs["input_ids"].squeeze(0),
            "rejected_attention_mask": rejected_inputs[
                "attention_mask"
            ].squeeze(0),
            "rejected_labels": rejected_inputs["input_ids"].squeeze(0).clone(),
        }

    def __len__(self):
        return len(self.data)
```

**Comment by yunsangju** [comment]
@eric-mitchell 
I made an issue. I'd appreciate it if you could check it out.

**Comment by eric-mitchell** [comment]
Thanks for sharing this. I'm not super familiar with pytorch lightning- out of curiosity, is there a reason you've refactored the trainer yourself?

**Comment by eric-mitchell** [comment]
@yunsangju just checking in- any progress on this issue?

**Comment by eric-mitchell** [comment]
@yunsangju I will close this for now, but please feel free to re-open if you have any new info or questions!
